Update command descriptions in README

adityasaky opened the pull request:
**Fixes issue**: The base-path argument was missing from the documentation for in-toto-run and in-toto-record in the README file.

**Description of the changes being introduced by the pull request**: The argument has been added to the relevant sections.

adityasaky committed:
Add base-path argument to in-toto-run and in-toto-record sections

lukpueh commented:
Thanks for the PR, @adityasaky! I just saw that ` base-path` is not the only missing argument.  And that the other commands in the README are out of sync with the implementation too.

What do you think about removing the detailed help message dumps from the README and just instruct the reader to use the respective command with the `--help` flag for more details?
I assume that the exact command line arguments are above all interesting for someone who is about to use the command and thus has it installed.

It would probably be more helpful to revise the prose descriptions of the commands in the README. Do you feel like doing that?

adityasaky commented:
> What do you think about removing the detailed help message dumps from the README and just instruct the reader to use the respective command with the --help flag for more details?

I think that's a good idea actually. The documentation is available via `--help` as you say and it's easier to maintain one set too. On the repo itself, we also have it in `runib.py` file under the respective function, so I think that's enough of a description too?

> It would probably be more helpful to revise the prose descriptions of the commands in the README. Do you feel like doing that?

I think I can after I work on the `lstrip` issue we discussed. Does that work?

lukpueh commented:
Yes, finishing #194 and then working on the README is a good plan. And yes, you can also reference to the source code documentation.

adityasaky commented:
I'm probably going to start work on updating README now. Is it okay if I stick to this branch and continue pushing here, so that it can be a part of this PR?

adityasaky committed:
Remove detailed commands and arguments from README and instead add --help

adityasaky committed:
Fix error with start of command block

adityasaky commented:
I've updated the sections for `in-toto-run` and `in-toto-record`. I'll do the same thing for `verify` and `sign` once I confirm with the respective source files.

adityasaky committed:
Remove detailed usage for in-toto-verify and in-toto-sign

lukpueh reviewed (changes requested):
Awesome! Here are a few suggestions:
- Writing "*command line arguments and usage examples*" instead of *"arguments and usage"* might be clearer
- Please wrap lines at 80 characters. I know most of the README does not follow that rule, but we should slowly enforce it when changing parts. It makes future diffs smaller.
- It's generally good (accessibility) practice to have link texts describe what they are linking to, e.g. 
 *"...[code documentation]()..."* instead of *"..documentation [here]()..."*.
- Would you mind taking a stab at the description text of `in-toto-run`? #234 supposedly addressed some suggestions made in #159. I think this could be better. Besides there is a mistake (*"...See **the [this** simple..."*) and I am not sure why it links to the "tampering with the software supply chain" part of the demo instead of the entire demo.

adityasaky commented:
I’ll get on those changes, @lukpueh, thanks!

adityasaky committed:
Fix line length in modified parts

adityasaky committed:
Make --help prompt text and link to documentation more verbose

adityasaky committed:
Fix grammatical error - 'the this'

adityasaky committed:
Update sample link to entire demo instead of subsection

adityasaky commented:
I fixed the first three issues, @lukpueh. Regarding the last one:

- For now, I fixed the grammatical error ("the this")
- I also updated the link to point to the demo instead of that particular subsection

When you say it can be better, do you mean a rewrite of that section? I'm looking over the initial threads that started this - #159 and comments stemming from #150.

lukpueh commented:
Yes, please take another look at #159, especially the first three bullet points. `in-toto-run` is one of the two most central tools of the in-toto framework (the other one is `in-toto-verify`). The README should summarize its purpose and usage and point to additional documentation (docstrings, cli help message, demo).

And yes, @vladimir-v-diaz's original comments in #150 and by @aaaaalbert [comment in #234](https://github.com/in-toto/in-toto/pull/234#issuecomment-454793725) may give you some ideas.

lukpueh commented on README.md:
```diff
@@ -115,24 +97,27 @@ Use `in-toto-verify` on the final product to verify that
+signed by multiple keys at once while link metadata can only be signed by one key at a time
+- write signed metadata to a specified path (if no output path is specified, 
+layout metadata is written to the path of the input file while link metadata 
+is written to '<name>.<keyid prefix>.link')
```
Please use `&lt;` and `&gt;` for  &lt;  and &gt; or put the whole filename inside backticks.

lukpueh commented on README.md:
```diff
@@ -115,24 +97,27 @@ Use `in-toto-verify` on the final product to verify that
-`in-toto-sign`, unlike `in-toto-verify`, can only add, replace, and verify signatures within in-toto Link or Layout metadata, with options to
-- replace (default) or add signature(s), with layout metadata able to be signed by multiple keys at once while link metadata can only be signed by one key at a time
-- write signed metadata to a specified path (if no output path is specified, layout metadata is written to the path of the input file while link metadata is written to '<name>.<keyid prefix>.link')
+`in-toto-sign`, unlike `in-toto-verify`, can only add, replace, and 
```
`in-toto-sign` is a *only* a metadata helper tool, whereas `in-toto-verify` is the command line interface for the entire in-toto verification routine. 

Please rephrase to something like ...
*`in-toto-sign` is a metadata signature helper tool to add, replace ....*

adityasaky replied:
Should the document outright say that `verify` is the command for the entire verification routine?

lukpueh replied:
No, I think that's not necessary.

adityasaky committed:
Add initial rewrite of in-toto-run adding more details about its function

adityasaky committed:
Switch lesser than / greater than symbols

adityasaky committed:
Clarify in-toto-sign's role as a helper tool

lukpueh commented on README.md:
```diff
@@ -66,19 +66,18 @@ To learn more about the different rule types, their guarantees and how they are
+which are files required for command execution, and products which are files
+expected as a result of the execution of the command. These, and other 
+relevant details pertaining to the step are stored in a link file, which is 
+signed using the functionary's key.
```
I think it could still be made a little clearer that `in-toto-run` should be used to executes any command corresponding to a step of the software supply chain, such as tagging a software release with git, running tests or building a binaries.

Also the explanation about materials might be a little confusing. Which command requires them for execution? `in-toto-run` or the passed command? And will execution fail if they are not passed? (*It does not, but the materials are simply not recorded in the corresponding link file, if not passed*).

lukpueh commented on README.md:
```diff
@@ -66,19 +66,18 @@ To learn more about the different rule types, their guarantees and how they are
+See [this simple usage example from the demo application 
+for more details](https://github.com/in-toto/demo). 
+For a detailed list of all the command line arguments, run `in-toto-run --help` 
+or look at the code documentation [here](https://github.com/in-toto/in-toto/blob/develop/in_toto/in_toto_run.py).
```
Please consider the comment I left about link tests (here and below).

adityasaky replied:
Do you mean the web layout wizard specified below?

adityasaky replied:
Oh do you mean this: 

> It's generally good (accessibility) practice to have link texts describe what they are linking to, e.g.
"...code documentation..." instead of "..documentation here...".

adityasaky replied:
Fixed. I was confused by tests.

lukpueh commented on README.md:
````diff
@@ -115,24 +104,27 @@ Use `in-toto-verify` on the final product to verify that
-                    [--gpg-home <path>] [-o <path>] [-a] [--verify] [-v | -q]
-```
+
+This tool is intended to sign layouts created by the layout web wizard, but 
````
Maybe we should link to the layout web wizard here:
https://in-toto.engineering.nyu.edu/

lukpueh reviewed (changes requested):
I like this a lot better. Can you give it another try and refine? Please also strip the trailing whitespace that you added.

adityasaky committed:
Resolve review - clarify regarding step command, and materials being optional

adityasaky committed:
Update links to contain their description in the text

adityasaky committed:
Add link to web layout creation wizard

adityasaky committed:
Remove accidental whitespace in in-toto-run section

lukpueh commented on README.md:
```diff
@@ -66,19 +66,20 @@ To learn more about the different rule types, their guarantees and how they are
+of the execution of that command. These, and other relevant details 
+pertaining to the step are stored in a link file, which is signed using the 
+functionary's key. If materials are not passed to the command, the link file 
+generated just doesn't record them.
```
Same goes for products.

lukpueh commented on README.md:
```diff
@@ -66,19 +66,20 @@ To learn more about the different rule types, their guarantees and how they are
+See [this simple usage example from the demo application 
+for more details](https://github.com/in-toto/demo). 
+For a detailed list of all the command line arguments, run `in-toto-run --help` 
+or look at the [code documentation](https://github.com/in-toto/in-toto/blob/develop/in_toto/in_toto_run.py) here.
```
I think you can remove the *"here"* now (here and below).

lukpueh reviewed (changes requested):
So I have two more small nitpicks in my inline comments.

And the trailing whitespace I mentioned in my last comment, was about the single trailing space on every line that you modified (see https://github.com/secure-systems-lab/code-style-guidelines/issues/5 for a discussion about it). But this is just a suggestion, not a blocker.

I'm also requesting a quick review from someone else, preferably not from the in-toto team. (@aaaaalbert? @alyptik?)

adityasaky committed:
Fix trailing white space on every line

adityasaky committed:
Remove 'here' from the documentation links

adityasaky committed:
Add clarification regarding recording of products

SantiagoTorres commented on README.md:
```diff
@@ -66,19 +66,22 @@ To learn more about the different rule types, their guarantees and how they are
-
+`in-toto-run` is used to execute a step in the software supply chain. This can be
+anything relevant to the project such as tagging a release with `git`, running
+a test, or building a binary to ship. The relevant step name and command are
```
I'd drop "to ship here"

SantiagoTorres commented on README.md:
```diff
@@ -115,24 +108,28 @@ Use `in-toto-verify` on the final product to verify that
+signed by multiple keys at once while link metadata can only be signed by one key at a time
+- write signed metadata to a specified path (if no output path is specified,
+layout metadata is written to the path of the input file while link metadata
+is written to '&lt;name&gt;.&lt;keyid prefix&gt;.link')
```
can' we change these inline html encoded <,> markers to actual symbols?

SantiagoTorres reviewed (changes requested):
Two very minor nits.

Thanks!

adityasaky committed:
Remove 'to ship' from the binary line

adityasaky committed:
Use '<>' symbols and switch to backticks

lukpueh commented on README.md:
```diff
@@ -66,19 +66,22 @@ To learn more about the different rule types, their guarantees and how they are
+functionary's key. If materials are not passed to the command, the link file
+generated just doesn't record them. Similarly, if the execution of a command
+via `in-toto-run` doesn't result in any products, they're not recorded in the
+link file.
```
IMHO the parts about materials and products are still ambiguous. The command reads/modifies/creates files no matter what the functionary passed as `--materials` or `--products`. And the reverse is true, in-toto records the hashes of passed `--materials` and `--products` no matter which files the command touches. So what is passed here is not immediately relevant for the command to execute, but it allows the functionary to explicitly provide information about the command execution, which in turn allows us to verify eventually that the steps of the supply chain are properly linked together. Does this make sense? Could you try once more to distill this information? :) Thanks!

adityasaky replied:
I'll give it a shot.

lukpueh commented on README.md:
````diff
@@ -87,19 +90,9 @@ by a single command. Use `in-toto-record start ...` to create a
-```
+link metadata file. For a detailed list of all command line arguments and their usage,
+run `in-toto-record start --help` or `in-toto-record stop --help`. You can also
+find the [code documentation](https://github.com/in-toto/in-toto/blob/develop/in_toto/in_toto_record.py).
````
This last sentence sounds a bit odd. Can you rephrase it to something like above or below.

adityasaky committed:
Modify code documentation link statement for in-toto-record

adityasaky committed:
Add further explanation regarding recording of artifacts during run

adityasaky committed:
Modify explanation regarding recording of artifacts during run

adityasaky committed:
Fix line length
